File: genform.php
<?php

interface FieldInterface {
    public function getTag(): string;
    public function getType(): string;
    public function getAttributes(): array;
    public function getText(): string;
    public function getOptions(): array;
}

class Field implements FieldInterface {
    private $tag, $type, $attributes, $text, $options;

    public function __construct(array $field_data) {
        foreach ($field_data as $prop => $value) {
            if (!property_exists($this, $prop)) {
                throw new Exception("Invalid property {$prop} on Field instance");
            } else {
                $this->{$prop} = $value;
            }
        }
        $this->_validate();
    }

    private function _validate(): void {
        foreach (get_class_vars(FieldInterface::class) as $prop => $value) {
            if (!property_exists($this, $prop)) {
                throw new Exception("{$prop} is required on Field instance");
            }
        }
    }

    public function getTag(): string {
        return $this->tag;
    }

    // And so on for other methods...
}

class FormGenerator {
    private $field_count = 0;
    private $form_container = [];

    public function addField(array $field_data): void {
        $field = new Field($field_data);
        $element = [];
        $tag = $field->getTag() ?? $field->getType();
        if (in_array($tag, ['span', 'select', 'textarea', 'label', 'input', 'radio', 'checkbox'])) {
            $element['tag'] = $tag;
            foreach ($field->getAttributes() as $attr => $value) {
                $element[$attr] = $value;
            }
            // And so on for other conditions...
        } else {
            throw new Exception("Unsupported tag {$tag}");
        }
    }

    // And so on for other methods...
}

?>

File: genform.py
from abc import ABC, abstractmethod

class FieldInterface(ABC):
    @property
    @abstractmethod
    def tag(self) -> str:
        pass

    @property
    @abstractmethod
    def type(self) -> str:
        pass

    @property
    @abstractmethod
    def attributes(self) -> dict:
        pass

    @property
    @abstractmethod
    def text(self) -> str:
        pass

    @property
    @abstractmethod
    def options(self) -> list:
        pass

class Field(FieldInterface):
    def __init__(self, field_data: dict):
        for prop in field_data:
            if not hasattr(self, prop):
                raise ValueError(f"Invalid property {prop} on Field instance")
            setattr(self, prop, field_data[prop])
        self._validate()

    def _validate(self):
        interface = FieldInterface.__dict__
        for prop in interface:
            if not hasattr(self, prop) and callable(getattr(FieldInterface, prop)):
                raise ValueError(f"{prop} is required on Field instance")

class FormGenerator:
    def __init__(self):
        self.field_count = 0
        self.form_container = []

    def add_field(self, field_data: dict) -> None:
        field = Field(field_data)
        element = {}
        tag = field.tag or field.type
        if tag in ['span', 'select', 'textarea', 'label', 'input', 'radio', 'checkbox']:
            element['tag'] = tag
            for attr, value in field.attributes.items():
                element[attr] = value
            if tag == 'span':
                element['text'] = field.text
            elif tag == 'select':
                element['options'] = [{'value': opt.get('attributes', {}).get('value'), 'text': opt.get('attributes', {}).get('text')} for opt in field.options]
        else:
            raise ValueError(f"Unsupported tag {tag}")

        if not element.get('name'):
            element['name'] = f'field{self.field_count}'
            self.field_count += 1

        self.form_container.append(element)
